avsomr/avsomr.h

```diff
@@ -0,0 +1,182 @@
+#ifndef AVS_AVSOMR_H
+#define AVS_AVSOMR_H
+
+#include <memory>
+#include <QList>
+#include <QHash>
+#include <QSet>
+#include <QRect>
+#include <QImage>
+#include <QByteArray>
+
+namespace Ms {
+namespace Avs {
+
+class MsmrFile;
+class AvsOmr
+      {
+   public:
+      AvsOmr();
+      using ID = uint32_t;
+      using Num = uint16_t;
+      using Idx = uint16_t;
+
+      // Model
+      struct Barline {
+            ID id{0};
+            ID glyphID{0};
+            };
+
+      struct Staff {
+            ID id{0};
+            QList<ID> barlines;
+
+            struct {
+                  int32_t start{0};
+                  int32_t stop{0};
+                  ID clefID{0};
+                  ID keyID{0};
+                  ID timeID{0};
+                  } header;
+
+            bool isValid() const { return id > 0; }
+            };
+
+      struct MStack {
+            ID id{0};
+            Idx idx{0}; // resolved
+            int32_t left{0};
+            int32_t right{0};
+            bool isValid() const { return right > 0; }
+            };
+
+      struct Part {
+            QList<Staff> staffs;
+            };
+
+      struct Inters {
+            QHash<ID, Barline> barlines;
+            QSet<ID> usedglyphs;
+            };
+
+      struct System {
+            QList<MStack> mstacks;
+            Part part;
+            Inters inters;
+            QSet<ID> freeglyphs;
+
+            // resolved
+            int32_t top{0};
+            int32_t bottom{0};
+
+            const MStack& stackByIdx(Idx idx, Idx* idxInSys = nullptr) const;
+            };
+
+      struct Page {
+            QList<System> systems;
+            };
+
+      enum class GlyphUsed {
+            Undefined = 0,
+            Used,
+            Free,
+            Free_Covered,   // Avs marked as free, but fully covered used
+            Trash           // Detected as trash (too small)
+            };
+
+      struct Glyph {
+            ID id{0};
+            QRect bbox;
+            GlyphUsed used{GlyphUsed::Undefined};
+            QImage img;
+            };
+
+      struct Sheet {
+            Num num{0};
+            Page page;
+            Idx mbeginIdx{0};
+            Idx mendIdx{0};
+            QHash<ID, Glyph*> glyphs;
+
+            bool isGlyphUsed(const ID& glypthID) const;
+            bool isGlyphFree(const ID& glypthID) const;
+            };
+
+      struct Book {
+            uint16_t sheets{0};
+            };
+
+      void resolve();
+
+      // Configure
+      struct Config {
+
+            void setIsHideAll(bool arg) {
+                  _isHideAll = arg;
+                  }
+            bool isHideAll() const { return _isHideAll; }
+
+            void setIsHideRecognized(bool arg) {
+                  _isHideRecognized = arg;
+                  }
+            bool isHideRecognized() const { return _isHideRecognized; }
```

Comment:
Hmm, you may want to put all the one-line functions each in one line...

Reply:
fixed